rust/lance-encoding/src/encodings/logical/primitive.rs

```diff
@@ -704,89 +1156,318 @@ impl MiniBlockScheduler {
-    // same as the sum of `Self::ranges`
-    vals_targeted: u64,
-    ranges: Vec<Range<u64>>,
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+enum PreambleAction {
+    Take,
+    Skip,
+    Absent,
 }
 
-impl Clone for ScheduledChunk {
-    fn clone(&self) -> Self {
-        Self {
-            data: self.data.try_clone().unwrap(),
-            vals_in_chunk: self.vals_in_chunk,
-            ranges: self.ranges.clone(),
-            vals_targeted: self.vals_targeted,
+// TODO: Add test cases for the all-preamble and all-trailer cases
+
+// When we schedule a chunk we use the repetition index (or, if none exists, just the # of items
+// in each chunk) to map a user requested range into a set of ChunkInstruction objects which tell
+// us how exactly to read from the chunk.
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct ChunkInstructions {
+    // The index of the chunk to read
+    chunk_idx: usize,
+    // A "preamble" is when a chunk begins with a continuation of the previous chunk's list.  If there
+    // is no repetition index there is never a preamble.
+    //
+    // It's possible for a chunk to be entirely premable.  For example, if there is a really large list
+    // that spans several chunks.
+    preamble: PreambleAction,
+    // How many complete rows (not including the preamble or trailer) to skip
+    //
+    // If this is non-zero then premable must not be Take
+    rows_to_skip: u64,
+    // How many complete (non-preamble / non-trailer) rows to take
+    rows_to_take: u64,
+    // A "trailer" is when a chunk ends with a partial list.  If there is no repetition index there is
+    // never a trailer.
+    //
+    // It's possible for a chunk to be entirely trailer.  This would mean the chunk starts with the beginning
+    // of a list and that list is continued in the next chunk.
+    //
+    // If this is true then we want to include the trailer
+    take_trailer: bool,
+}
+
+// First, we schedule a bunch of [`ChunkInstructions`] based on the users ranges.  Then we
+// start decoding them, based on a batch size, which might not align with what we scheduled.
+//
+// This results in `ChunkDrainInstructions` which targets a contiguous slice of a `ChunkInstructions`
+//
+// So if `ChunkInstructions` is "skip preamble, skip 10, take 50, take trailer" and we are decoding in
+// batches of size 10 we might have a `ChunkDrainInstructions` that targets that chunk and has its own
+// skip of 17 and take of 10.  This would mean we decode the chunk, skip the preamble and 27 rows, and
+// then take 10 rows.
+//
+// One very confusing bit is that `rows_to_take` includes the trailer.  So if we have:
+// "no preamble, skip 5, take 10, take trailer" and we are draining 20 rows then the
+// first drain instructions will have no preamble, skip 0, take 11 and the second chunk
+// instructions will have take preamble, skip 0, take 9 (assuming the second chunk has at
+// least 9 rows)
```

Comment:
Do we ignore the skips? Also why 11 when the total is take 10? Is that because the trailer is counting as 1 here?

Reply:
Maybe this could make more sense in some other way but in my head:

The chunk instructions describe a range into the chunk.  The drain instructions describe a range into the chunk instructions:

![image](https://github.com/user-attachments/assets/01984298-2a7a-4c8c-b10c-d21c686a8ee0)

Reply:
And yes, the total is 11 because of the trailer.  When we're acutally zoomed into the rep levels and mapping the ranges the trailer looks like a normal row (e.g. look at the middle chunk below):

```
1 0 0 0 1 0 | 0 0 1 0 0 1 0 0 1 0 0 1 0 0 | 0 0 1 0
```

So the logic was (at the time) simpler to just treat it as a row by the time we hit map_ranges.  Maybe it would be simpler if we keep it a separate boolean?
```

Reply:
I don't think we need a separate boolean, just wanted to clarify the logic here. Also TBH I'm confused about the concept of draining 20 rows from a take 10.

Reply:
It was meant to be draining 20 rows from a series of chunks starting with a chunk that only has 10 rows and a trailer.  However, I cleaned up the wording and made the example more explicit:

```
// One very confusing bit is that `rows_to_take` includes the trailer.  So if we have two chunks:
//  -no preamble, skip 5, take 10, take trailer
//  -take preamble, skip 0, take 50, no trailer
//
// and we are draining 20 rows then the drain instructions for the first batch will be:
//  - no preamble, skip 0 (from chunk 0), take 11 (from chunk 0)
//  - take preamble (from chunk 1), skip 0 (from chunk 1), take 9 (from chunk 1)
```